actions.py

```diff
@@ -0,0 +1,62 @@
+import pickle
+import base64
+from pathlib import Path
+import subprocess
+
+from flask import Blueprint, request, jsonify, session
+
+bp = Blueprint("actions", __name__)
+
+
+@bp.route("/message", methods=["POST"])
+def log_entry():
+    user_info = session.get("user_info", None)
+    if user_info is None:
+        return jsonify({"error": "no user_info found in session"})
+    access_level = user_info[2]
+    if access_level > 2:
+        return jsonify({"error": "access level < 2 is required for this action"})
+    filename_param = request.form.get("filename")
+    if filename_param is None:
+        return jsonify({"error": "filename parameter is required"})
+    text_param = request.form.get("text")
+    if text_param is None:
+        return jsonify({"error": "text parameter is required"})
+
+    user_id = user_info[0]
+    user_dir = "data/" + str(user_id)
+    user_dir_path = Path(user_dir)
+    if not user_dir_path.exists():
+        user_dir_path.mkdir()
+
+    filename = filename_param + ".txt"
+    path = Path(user_dir + "/" + filename)
+    # vulnerability: Directory Traversal
+    with path.open("w", encoding="utf-8") as open_file:
+        open_file.write(text_param)
+    return jsonify({"success": True})
+
+
+@bp.route("/grep_processes")
+def grep_processes():
+    name = request.args.get("name")
+    # vulnerability: Remote Code Execution
+    res = subprocess.run(
+        ["ps aux | grep " + name + " | awk '{print $11}'"],
+        shell=True,
+        capture_output=True,
+    )
+    if res.stdout is None:
+        return jsonify({"error": "no stdout returned"})
+    out = res.stdout.decode("utf-8")
+    names = out.split("\n")
+    return jsonify({"success": True, "names": names})
+
+
+@bp.route("/deserialized_descr", methods=["POST"])
+def deserialized_descr():
+    pickled = request.form.get('pickled')
+    data = base64.urlsafe_b64decode(pickled)
+    # vulnerability: Insecure Deserialization
+    deserialized = pickle.loads(data)
```

Comment:
[comment]: <> (FrogbotReview Sast 61 20 61 38 pickle.loads\(data\) /actions.py)

### 🔐 Static Application Security Testing (SAST) Vulnerability 
	
| Severity | Finding |
| :---: | :---: |
| ![](https://raw.githubusercontent.com/jfrog/frogbot/master/resources/v2/applicableHighSeverity.png)<br> | Unsafe Deserialization |

#### 👇 Details

<details>
<summary> <b>Full description</b> </summary>
<br>


### Overview
Unsafe deserialization is a security vulnerability that occurs when a program deserializes
untrusted data with a potentially dangerous deserializer.
Deserialization is the process of converting serialized data (data that
has been converted into a format that can be easily transmitted or stored) back into its
original form. In some ("unsafe") serialization protocols, if an attacker is able to
manipulate the serialized data, they may be able to execute arbitrary code or perform other
malicious actions when the data is deserialized.

### Query operation
In this query we look for user input that can flow un-sanitized to potentially unsafe
deserialization methods

### Vulnerable example
```python
import yaml
from flask import Flask, request

app = Flask(__name__)

@app.route('/process', methods=['POST'])
def process():
    data = request.get_data()

    # Vulnerable deserialization
    obj = yaml.load(data)

    # Process the deserialized object (for simplicity, we're just printing it)
    print(obj)

    return 'Data processed'

if __name__ == '__main__':
    app.run()
```
In this example, the application exposes a `/process` endpoint that accepts data via a POST
request. The vulnerable code uses the `yaml.load()` function
to deserialize the received data.
The vulnerability lies in the fact that the `yaml` module can execute arbitrary code
during the deserialization process.
An attacker can exploit this by crafting a malicious payload
that executes arbitrary code when the `yaml.load()` function is called.

### Remediation
Use deserialization routines that are known to handle untrusted data securely, such as
`yaml.safe_load`. It is highly recommended to use the `json` module for serialization, as it
deserializes untrusted data securely.

```diff
@app.route('/process', methods=['POST'])
def process():
    data = request.get_data()

    # Safe deserialization
-    obj = yaml.load(data)
+    obj = yaml.safe_load(data)

    # Process the deserialized object (for simplicity, we're just printing it)
    print(obj)

    return 'Data processed'
```


</details>



<details>
<summary> <b>1. Vulnerable data flow analysis result</b> </summary>
<br>

1. `request.form` (at /actions.py line 58)

2. `request.form.get` (at /actions.py line 58)

3. `request.form.get('pickled')` (at /actions.py line 58)

4. `pickled` (at /actions.py line 58)

5. `base64.urlsafe_b64decode(pickled)` (at /actions.py line 59)

6. `data` (at /actions.py line 59)

7. `pickle.loads(data)` (at /actions.py line 61)


</details>


---
<div align="center">

[🐸 JFrog Frogbot](https://github.com/jfrog/frogbot#readme)

</div>